File: 2022/07/16/java/IntersectionTwoArrays.java
/**
 * https://leetcode.com/problems/intersection-of-two-arrays/
 * @author Andrew Jarombek
 * @since 7/16/2022
 */

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class IntersectionTwoArrays {
    public static int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set = new HashSet<>();

        for (int num : nums1) {
            set.add(num);
        }

        List<Integer> list = new ArrayList<>();

        for (int num : nums2) {
            if (set.contains(num)) {
                list.add(num);
                set.remove(num);
            }
        }

        int[] result = new int[list.size()];

        for (int i = 0; i < result.length; i++) {
            result[i] = list.get(i);
        }

        return result;
    }

    public static void main(String[] args) {
        int[] nums1 = new int[] {1, 2, 2, 1};
        int[] nums2 = new int[] {2, 2};
        int[] result = intersection(nums1, nums2);

        assert result.length == 1;
        assert result[0] == 2;
    }
}

File: 2023/02/12/go/intersection_two_arrays.go
/**
 * https://leetcode.com/problems/intersection-of-two-arrays/
 * Author: Andrew Jarombek
 * Date: 2/12/2023
 */

package _go

func intersection(nums1 []int, nums2 []int) []int {
	set1 := make(map[int]int)
	set2 := make(map[int]int)

	for _, num := range nums1 {
		set1[num] = 1
	}

	for _, num := range nums2 {
		set2[num] = 1
	}

	result := []int{}

	for key, _ := range set1 {
		if set2[key] == 1 {
			result = append(result, key)
		}
	}

	return result
}
